### PR TITLE
Feature/cluster loss

### PR DESCRIPTION
**Pull request by julia-shenshina**
## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/contrib/nn/criterion/cluster.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            embeddings (torch.Tensor): tensor of shape (batch_size, embed_dim)
+            targets: labels of the batch, shape (batch_size,)
+
+        Returns:
+            torch.Tensor: cluster loss for the batch
+        """
+        return self._batch_hard_cluster_loss(embeddings, targets)
+
+
+__all__ = ["ClusterLoss"]
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/cluster.py]: could you please also add it to `catalyst/contrib/nn/criterion/__init__.py`?

Comment (Scitator) [reply catalyst/contrib/nn/criterion/cluster.py]: and to our docs here, https://github.com/catalyst-team/catalyst/blob/master/docs/api/contrib.rst#criterion

File: catalyst/contrib/nn/criterion/cluster.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        Forward propagation method for cluster loss.
+
+        Args:
+            embeddings (torch.Tensor): tensor of shape (batch_size, embed_dim)
+            targets: labels of the batch, shape (batch_size,)
+
+        Returns:
+            torch.Tensor: cluster loss for the batch
+        """
+        return self._batch_hard_cluster_loss(embeddings, targets)
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/cluster.py]: does it work with `targets` of shape `[bs; 1]`?

File: catalyst/contrib/nn/criterion/cluster.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+from torch import nn
+
+from catalyst.contrib.nn.criterion.functional import euclidean_distance
+
+
+class ClusterLoss(nn.Module):
+    """
+    Cluster loss
+    .. _Cluster Loss for Person Re-Identification
+    https://arxiv.org/pdf/1812.10325.pdf
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/cluster.py]: is it possible to add minimal example for this loss?
in long term it's much easier to understand them

something like
```
embeddings = toch.sample
targets = torch.sample
criterion = ClusterLoss()
loss = criterion(embeddings, targets)
```

Comment (Scitator) [reply catalyst/contrib/nn/criterion/cluster.py]: it would be also nice to see some kind of tests for this criterion like
https://github.com/catalyst-team/catalyst/blob/master/catalyst/utils/metrics/tests/test_iou.py
to ensure it correctness :)

File: catalyst/data/sampler_inbatch.py [hunk 4]
```diff
@@ -18,7 +19,39 @@
...
 
 
-class InBatchTripletsSampler(ABC):
+class IInbatchTripletSampler(ABC):
+    """
+    An abstraction of inbatch triplet sampler.
+    """
+
+    @abstractmethod
+    def _check_input_labels(self, labels: List[int]) -> None:
```
Comment (Scitator) [inline catalyst/data/sampler_inbatch.py:29]: do we really need this method?

File: catalyst/data/sampler_inbatch.py [hunk 5]
```diff
@@ -18,7 +19,39 @@
...
+        """
+        Check if the batch labels list is valid for the sampler.
+
+        Args:
+            labels: labels of the samples in the batch
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def sample(self, features: Tensor, labels: List[int]) -> TTriplets:
```
Comment (Scitator) [inline catalyst/data/sampler_inbatch.py]: `labels` should be also Tensor, isn't it? with shape of `[bs; ]`?

Comment (Scitator) [reply catalyst/data/sampler_inbatch.py]: just suggesting to add this to the docs :)

File: catalyst/data/sampler_inbatch.py [hunk 6]
```diff
@@ -31,12 +64,10 @@ class InBatchTripletsSampler(ABC):
     But you are not limited to using it in any other way.
     """
 
-    @staticmethod
```
Comment (Scitator) [inline catalyst/data/sampler_inbatch.py:36]: removed, why?

File: catalyst/data/sampler_inbatch.py [hunk 7]
```diff
@@ -195,3 +226,158 @@ def _sample_from_distmat(
...
         return ids_anchor, ids_pos, ids_neg
+
+
+class HardClusterSampler(IInbatchTripletSampler):
+    """
+    This sampler selects hardest triplets based on distance to mean vectors.
+    The batch must contain k samples for p classes in it, k > 1, p > 1.
+    """
+
+    def _check_input_labels(self, labels: List[int]) -> None:
```
Comment (Scitator) [inline catalyst/data/sampler_inbatch.py:256]: could we make it static?

File: catalyst/data/tests/test_sampler_inbatch.py [hunk 8]
```diff
@@ -236,3 +237,175 @@ def test_hard_sampler_manual() -> None:
...
+        mean_vectors: tensor of shape (p, embed_dim) -- mean vectors of
+        classes in the batch
+        expected: tensor of shape (p, p) -- expected distances from mean
+        vectors of classes
+    """
+    sampler = HardClusterSampler()
+    distances = sampler._count_inter_class_distances(  # noqa: WPS437
+        mean_vectors
+    )
+    print(distances)
```
Comment (Scitator) [inline catalyst/data/tests/test_sampler_inbatch.py]: do we need print here?

File: catalyst/data/sampler_inbatch.py (hunk 5 above)
Comment (Scitator) [reply catalyst/data/sampler_inbatch.py]: @julia-shenshina ping :)

File: catalyst/data/sampler_inbatch.py (hunk 4 above)
Comment (Scitator) [reply catalyst/data/sampler_inbatch.py:29]: @julia-shenshina and here :)

File: catalyst/data/utils.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+from typing import List, Union
+
+from torch import int as tint, long, short, Tensor
+
+
+def prepare_labels(labels: Union[Tensor, List[int]]) -> List[int]:
```
Comment (Scitator) [inline catalyst/data/utils.py]: could we rename it to `process_labels`? (we have some internal notation guide :) )

File: catalyst/data/sampler_inbatch.py [hunk 10]
```diff
@@ -200,3 +241,158 @@ def _sample_from_distmat(
             ids_neg.append(i_neg)
 
         return ids_anchor, ids_pos, ids_neg
+
+
+class HardClusterSampler(IInbatchTripletSampler):
```
Comment (Scitator) [inline catalyst/data/sampler_inbatch.py:245]: could we insert it to our metric learning test? what do you think about it?

File: catalyst/data/utils.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
+from typing import List, Union
+
+from torch import int as tint, long, short, Tensor
+
+
+def process_labels(labels: Union[Tensor, List[int]]) -> List[int]:
```
Comment (AlekseySh) [inline catalyst/data/utils.py]: `process_integers` sounds more general

File: catalyst/data/sampler_inbatch.py [hunk 12]
```diff
@@ -200,3 +241,164 @@ def _sample_from_distmat(
...
+        and choose the hardest positive sample from the batch and the hardest
+        mean vector for it.
+
+        Args:
+            features: tensor of shape (batch_size; embed_dim) that contains
+                k samples for each of p classes
+            labels: labels of the batch, list or tensor of size (batch_size,)
+
+        Returns:
+            triplet of (mean_vector, positive, negative_mean_vector)
```
Comment (AlekseySh) [inline catalyst/data/sampler_inbatch.py]: Please, add info that size of output is `p`

File: catalyst/data/sampler_inbatch.py [hunk 13]
```diff
@@ -200,3 +241,164 @@ def _sample_from_distmat(
...
+        indices = torch.diag(torch.ones(p)).type(torch.bool)
+        matrix[indices] = value
+        return matrix
+
+    def sample(self, features: Tensor, labels: TLabels) -> TTriplets:
+        """
+        This method selects the hardest positive and negative example for
+        each label in the batch. It counts mean vectors for all the labels
+        and choose the hardest positive sample from the batch and the hardest
+        mean vector for it.
```
Comment (AlekseySh) [inline catalyst/data/sampler_inbatch.py]: In comparison with docs from the top of the current class this one looks confused / ambiguous.
I offer to left only 1st exploration from the top of the class, it is really great.

```
    This sampler selects hardest triplets based on distance to mean vectors:
    anchor is a mean vector of features of i-th class in the batch,
    the hardest positive sample is the most distant from anchor sample of
    anchor's class, the hardest negative sample is the closest mean vector
    of another classes.
    The batch must contain k samples for p classes in it (k > 1, p > 1).
``` 
cc

Comment (AlekseySh) [reply catalyst/data/sampler_inbatch.py]: You can put this 1st peace of documentation here (in sample method) or in the top of the class.

**Review by AlekseySh — DISMISSED** [review]
Looks greate!

Please,
1. update `LossWithTripletsSampling`:
    - update typing of `init`
    - use `process_labels` instead of `prepare_labels`, because now we have duplicated code

2. add test which we have discussed in slack: `set(hard_tri()._sample()) == set(hard_clusters.sample())`

File: catalyst/data/sampler_inbatch.py [hunk 14]
```diff
@@ -10,15 +10,56 @@
 import torch
 from torch import Tensor
 
+from catalyst.contrib.nn.criterion.functional import euclidean_distance
```
Comment (AlekseySh) [inline catalyst/data/sampler_inbatch.py]: you can also use `torch.pdist()`

File: catalyst/data/utils.py (hunk 11 above)
Comment (Scitator) [reply catalyst/data/utils.py]: as far as this is function on top of `labels` name like `somethin_labels` sounds great
btw, could we name it to `convert_labels2list`? because, it looks like it's exactly what this function is doing :)
